Cython/Compiler/FusedNode.py

```diff
@@ -220,6 +220,10 @@ def _specialize_function_args(self, args, fused_to_specific):
                 arg.type = arg.type.specialize(fused_to_specific)
                 if arg.type.is_memoryviewslice:
                     arg.type.validate_memslice_dtype(arg.pos)
+                if arg.annotation:
+                    # TODO might be nice if annotations were specialized instead?
+                    # (Or might be hard to do reliably)
```

Comment:
Re-evaluation of an annotation will give the fused type (again). However, in a fused function that ends up overwriting the specialized type. Originally I'd tried to do it by looking if the annotation gave was a subtype of the fused type.
https://github.com/cython/cython/blob/6a30fecff5decdf20029763afea6183de3177dc3/Cython/Compiler/Nodes.py#L1683
The trouble with that is that it breaks for Python objects (since the assigned type starts as a Python object so the annotation never gets applied).

I don't think we can just ignore annotations in a fused specialization because I don't think the function knows its a fused specialization.

We can't use use `fused_to_specific` because that's applied after `analyse_declarations` https://github.com/cython/cython/blob/daf7e9534021e513a1c7da90732bb1e1b9638671/Cython/Compiler/FusedNode.py#L103-L106 and is also seems to be explicitly turned off when analysing the argument types https://github.com/cython/cython/blob/daf7e9534021e513a1c7da90732bb1e1b9638671/Cython/Compiler/Nodes.py#L2920

It definitely seems like getting types from annotations could be done better (i.e. in one place) - they're a bit scattered all over the place. That's obviously a bigger change

Reply:
It fixes a bug, let's keep it like it is for now.